009-crd-v1-roadmap.md

```diff
@@ -0,0 +1,108 @@
+
+
+# Roadmap to using the CRD v1 API
+
+This describes (roughly) the steps Strimzi users will need to take to get to using v1beta2 Strimzi APIs which, in turn, are using the v1 CRD API. In it we talk about Strimzi 0.21.0 and 0.22.0, but the changes described under those headings wouldn’t necessarily need to be delivered in exactly those versions. Rather the requirement is that what we’re calling Strimzi 0.22 is released around the time that Kubernetes 1.21 is.
+
+## Current situation
+
+Currently Strimzi defines numerous CustomResourceDefinitions) using the `apiextensions.k8s.io/v1beta1`. Those CRDs define the "Strimzi CRs", as v1alpha1 and v1beta1 APIS in the `kafka.strimzi.io` API group.
+
+Kubernetes 1.21 is expected to have removed CRD `apiextensions.k8s.io/v1beta1`. Its replacement, `apiextensions.k8s.io/v11`, has been available since Kubernetes 1.16.
+
+
+## Motivation
+
+To work on Kubernetes 1.21 Strimzi will need to use the v1 CRD API. We can't have a single set of CRDs which support both Kubernetes 1.21 or later and also 1.15 or earlier. Furthermore, using the v1 CRD API requires that the valiation schemas in the CRD are all "structural schemas".
+
+## Proposal
+
+### Strimzi 0.20
+
+This will be the last version of Strimzi which supports Kubernetes 1.11-1.15. 
+
+#### Users
+
+Users should:
+
+ * Rewrite their Kafka `spec.kafka.listeners` fields to use a list. 
+ * Check they’re not using any “deprecated” fields in any of their resources.
+
+### Strimzi "0.21.0"
+
+This release will be a gate release, meaning that all existing Strimzi users will have to pass through 0.21 in order to upgrade to 0.22.
+This should work on Kubernetes 1.16-1.20. It can’t work on lower versions because it relies on having support for per-versions validation schemas in CRDs, a feature that wasn’t available before 1.16. It can’t work on higher versions because that don’t support `v1beta1` of the CRD API.
+
+#### Development
+
+We need to:
+
+* Audit all existing CRD looking for things we need to change in order to have a structural schema. This is basically about checking that the type of each property is not going to need to change.
+* Prime suspects include things which are declared as a `Map<String, Object>` is Java, hence an object in JSON Schema. This includes `Kafka.spec.kafka.jvmOptions`, `Kafka.spec.zookeeper.jvmOptions`, and similar in other CRDs.
+For each of these we should make the analogous change that we did for listeners in 0.20.0.
```

Comment:
My feeling is that we should hurry up and change this in 0.20.0 or in some version before the v1beta2 move and not in the same version. I guess technically it is not critical, but it will make it IMHO smoother.

Reply:
I agree:

* technically it doesn't matter
* we should work on this as a matter of priority